src/storage/src/source/persist_source.rs

```diff
@@ -146,13 +166,27 @@ where
             .await
             .expect("cannot serve requested as_of");
 
-        loop {
+        let mut done = false;
+        while !done {
             while let Ok(leased_part) = consumed_part_rx.try_recv() {
                 subscription.return_leased_part(leased_part.into());
             }
 
-            yield subscription.next().await;
+            let (parts, progress) = subscription.next().await;
+            if timely::PartialOrder::less_equal(&until_clone, &progress) {
```

Comment:
Is this correct for partially ordered times? (which yes, we don't have right now 😅)

Would `!timely::PartialOrder::less_than(&progress, &until_clone)` be the correct one to use in that case?

Reply:
I think the above is correct. If `until <= progress` it means that all future batches will have timestamps that are greater or equal to `until`, which means they can be suppressed in their entirety. Nothing to do with partial or total orders, as far as I can tell. But push back if you don't buy that; it's easy for me to get myself turned around.

Reply:
I believe you! I first had to debug my mistake, though: I was wrongly thinking of `Antichain::less_equal(ts)` where it's enough for any element of the antichain to be <= the given timestamp. Just now, I thought this is also true for `PartialOrder::less_equal(&self, &other)` on Antichains, which it is not. 😅 We need to have an element in self that is <= x for all x in other.